File: sdk-python/src/airprompter_agent/sync/loop.py
"""One sync pass, and the three ways of scheduling it.

::

    resident:   every poll_seconds (jittered) — the edge pointer first,
                the manifest only when the generation moved
    on_invoke:  the same pass at invocation start and end (serverless has
                no background timer)
    daemon:     delegate to the host's daemon socket when present (T26);
                until then, in-process

A pass never blocks a render and never raises past its caller: sync
failures degrade to the last verified release and are reported. Every
manifest goes through the trust chain before a byte is staged; payloads
already held for unchanged hashes are reused, so a pass fetches only what
moved.
"""

from __future__ import annotations

import random
from dataclasses import dataclass
from typing import Any, Callable, Mapping, Optional, Sequence

from ..protocol.trust import referenced_payloads, verify_manifest, verify_root_metadata
from ..store.slot_store import LoadedSlot, SlotStore
from .client import SyncClient

#: ``activated_externally``: the customer's hook (or an operator) already made the staged release live during the decision.
ApplyPolicyDecision = str  # "activated" | "staged" | "activated_externally"


@dataclass
class SyncPassOutput:
    #: ``pointer_unchanged``: the unsigned edge pointer said nothing moved — silence, not contact (S3).
    #: ``unchanged``: the origin's authenticated 304, or a signed manifest at the generation already held — contact.
    outcome: str  # "pointer_unchanged" | "unchanged" | "activated" | "activated_externally" | "staged" | "refused" | "unavailable" | "nothing_promoted" | "held_back"
    etag: Optional[str]
    edge_etag: Optional[str]
    trusted_root: Mapping[str, Any]
    active: Optional[LoadedSlot]
    generation: Optional[int] = None
    reason: Optional[str] = None


def required_models_missing(payload: Mapping[str, Any], catalog: Optional[Sequence[str]]) -> list[str]:
    """The required models of a payload (its slots and every arm override) the declared catalog lacks; empty when nothing was declared."""
    if catalog is None:
        return []
    declared = set(catalog)
    slots = list(payload.get("slots", []))
    for arm in (payload.get("experiment") or {}).get("arms", []):
        slots.extend(arm.get("overrides", []))
    return sorted({slot["model"] for slot in slots if slot.get("modelRequired") is True and slot["model"] not in declared})


def sync_once(
    *,
    store: SlotStore,
    client: SyncClient,
    now: Callable[[], str],
    scope: Mapping[str, str],
    trusted_root: Mapping[str, Any],
    active: Optional[LoadedSlot],
    etag: Optional[str],
    apply_policy: Callable[[Mapping[str, Any]], ApplyPolicyDecision],
    fetch_root: Optional[Callable[[], Optional[Mapping[str, Any]]]] = None,
    edge_pointer_url: Optional[str] = None,
    edge_etag: Optional[str] = None,
    skip_pointer: bool = False,
    require_countersign: Optional[bool] = None,
    countersign_root: Optional[Mapping[str, Any]] = None,
    on_refusal: Optional[Callable[[str, Optional[int]], None]] = None,
    on_directives: Optional[Callable[[Mapping[str, Any]], None]] = None,
    catalog: Optional[Sequence[str]] = None,
    on_model_unavailable: Optional[Callable[[list[str], int], None]] = None,
) -> SyncPassOutput:
    """``on_directives`` (T9) is called with every manifest whose envelope verifies BEFORE the pass decides whether to stage,
    hold back or ignore it — so a ``disable`` (Freeze) or a ``request_unlock`` rides a manifest the runtime would otherwise
    leave staged or unchanged. Never a manifest that failed the trust chain."""
    at = now()
    root = trusted_root
    current_edge_etag = edge_etag

    def done(outcome: str, *, active_slot: Optional[LoadedSlot] = active, next_etag: Optional[str] = etag, generation: Optional[int] = None, reason: Optional[str] = None) -> SyncPassOutput:
        return SyncPassOutput(outcome, next_etag, current_edge_etag, root, active_slot, generation, reason)

    def refuse(reason: str, generation: Optional[int]) -> None:
        if on_refusal:
            on_refusal(reason, generation)

    try:
        # A newer root document is accepted only against the one already trusted (R1–R5).
        if fetch_root:
            candidate = fetch_root()
            if candidate:
                verdict = verify_root_metadata(candidate=candidate, trusted=root, now=at)
                if verdict.ok:
                    root = candidate
                    store.accept_root(candidate)
                else:
                    refuse(verdict.reason or "root_signature_invalid", None)

        # Idle path: the edge pointer says whether anything moved, without a Lambda on the other end. It is unsigned and
        # cacheable, so its silence is never contact (S3): the lease does not move on ``pointer_unchanged``.
        if edge_pointer_url and not skip_pointer:
            edge = client.edge_pointer(edge_pointer_url, current_edge_etag)
            if edge.status == "not_modified":
                return done("pointer_unchanged")
            if edge.status == "ok" and edge.pointer is not None:
                current_edge_etag = edge.etag
                if active and edge.pointer.get("generation", 0) <= active.generation:
                    return done("pointer_unchanged")

        fetched = client.manifest(if_none_match=etag)
        if fetched.status == "not_modified":
            return done("unchanged")
        if fetched.status == "not_found":
            return done("nothing_promoted")
        if fetched.status in ("unauthorized", "forbidden"):
            refuse(fetched.status, None)
            return done("unavailable", reason=fetched.status)
        if fetched.status == "error":
            return done("unavailable", reason=f"http_{fetched.http_status}")

        manifest = fetched.manifest or {}
        generation = manifest["payload"]["generation"]
        stored = store.state["generation"]
        envelope = verify_manifest(manifest=manifest, root=root, now=at, scope=scope, stored_generation=stored, payloads=None, countersign_root=countersign_root, require_countersign=require_countersign)
        if not envelope.ok:
            refuse(envelope.reason or "signature_invalid", generation)
            return done("refused", reason=envelope.reason, generation=generation)
        # The signature verified and the generation is not a rollback: its directives stand from here on.
        if on_directives:
            on_directives(manifest["payload"])
        if generation == stored:
            return done("unchanged", next_etag=fetched.etag)
        held_back_below = store.state.get("heldBackBelow")
        if held_back_below is not None and generation <= held_back_below:
            # A local rollback stepped down from this generation on purpose; only a newer one ends the hold.
            return done("held_back", next_etag=fetched.etag, generation=generation)
        # T15: a required model this runtime cannot call refuses the release here — verified, never fetched, never staged.
        missing_models = required_models_missing(manifest["payload"], catalog)
        if missing_models:
            if on_model_unavailable:
                on_model_unavailable(missing_models, generation)
            refuse("model_unavailable", generation)
            return done("refused", reason="model_unavailable", next_etag=fetched.etag, generation=generation)

        # Fetch only what moved; the bytes already verified in the active slot are reused for unchanged hashes.
        payloads: dict[str, bytes] = {}
        for content_hash in referenced_payloads(manifest["payload"]):
            held = active.payloads.get(content_hash) if active else None
            if held is not None:
                payloads[content_hash] = held
                continue
            data = client.payload(content_hash)
            if data is None:
                refuse("payload_missing", generation)
                return done("refused", reason="payload_missing", generation=generation)
            payloads[content_hash] = data
        full = verify_manifest(manifest=manifest, root=root, now=at, scope=scope, stored_generation=stored, payloads=payloads, countersign_root=countersign_root, require_countersign=require_countersign)
        if not full.ok:
            refuse(full.reason or "signature_invalid", generation)
            return done("refused", reason=full.reason, generation=generation)

        # All-or-nothing: the current slot stays whole until the new one is complete and fsynced.
        store.stage(manifest=manifest, payloads=payloads)
        decision = apply_policy(manifest)
        if decision == "staged":
            return done("staged", next_etag=fetched.etag, generation=generation)
        # The hook activated it itself: the caller's active slot already moved; nothing here to load.
        if decision == "activated_externally":
            return done("activated_externally", next_etag=fetched.etag, generation=generation)
        slot = store.activate()
        loaded = store.load(slot, now=at, root=root, countersign_root=countersign_root, require_countersign=require_countersign)
        return done("activated", active_slot=loaded, next_etag=fetched.etag, generation=generation)
    except Exception as error:  # noqa: BLE001 — a sync pass never raises past its caller
        refuse(f"network:{error}", None)
        return done("unavailable", reason="network")


def jittered_delay_ms(base_seconds: float, rand: Callable[[], float] = random.random) -> int:
    """±20 %: a fleet restarted together must not poll together."""
    jitter = (rand() * 2 - 1) * 0.2
    return max(1000, int(round(base_seconds * 1000 * (1 + jitter))))

File: sdk-typescript/packages/sync/src/sync/loop.ts
/**
 * One sync pass, and the three ways of scheduling it.
 *
 *   resident:   every pollSeconds (jittered) — the edge pointer first,
 *               the manifest only when the generation moved
 *   on_invoke:  the same pass at invocation start and end (serverless has
 *               no background timer)
 *   daemon:     delegate to the host's daemon socket when present (T26);
 *               until then, in-process
 *
 * A pass never blocks a render and never throws past its caller: sync
 * failures degrade to the last verified release and are reported. Every
 * manifest goes through the trust chain before a byte is staged; payloads
 * already held for unchanged hashes are reused, so a pass fetches only what
 * moved.
 */

import { referencedPayloads, verifyManifest, verifyRootMetadata } from "@airprompter/agent-core";
import type { Manifest, RefusalCode, RootMetadata } from "@airprompter/agent-core";
import type { LoadedSlot, SlotStore } from "../store/slotStore.js";
import type { SyncClient } from "@airprompter/agent-core";

/** `activated_externally`: the customer's hook (or an operator) already made the staged release live during the decision. */
export type ApplyPolicyDecision = "activated" | "staged" | "activated_externally";

export interface SyncPassResult {
  /**
   * `pointer_unchanged`: the unsigned edge pointer said nothing moved — silence, not contact (S3).
   * `unchanged`: the origin's authenticated `304`, or a signed manifest at the generation already held — contact.
   */
  outcome: "pointer_unchanged" | "unchanged" | "activated" | "activated_externally" | "staged" | "refused" | "unavailable" | "nothing_promoted" | "held_back";
  generation?: number;
  reason?: RefusalCode | "unauthorized" | "forbidden" | "network" | string;
}

export interface SyncPassInput {
  store: SlotStore;
  client: SyncClient;
  now: () => string;
  scope: { organizationId: string; agentId: string; target: "dev" | "staging" | "prod" };
  /** The last accepted root, or the synthetic pinned document. */
  trustedRoot: RootMetadata;
  /** S3: the heartbeat said the origin is ahead of what the pointer showed — go to the signed manifest, skip the pointer. */
  skipPointer?: boolean;
  /** A candidate root document fetched beside the manifest, when the runtime polls one. */
  fetchRoot?: () => Promise<RootMetadata | null>;
  /** What the active slot holds (its payload bytes are reused for unchanged hashes). */
  active: LoadedSlot | null;
  etag: string | null;
  edgePointerUrl?: string | null;
  edgeEtag?: string | null;
  requireCountersign?: boolean;
  countersignRoot?: RootMetadata | null;
  /** Local policy: `auto` activates a verified release; `unlock_required` stages it and calls `onStaged`. */
  applyPolicy: (manifest: Manifest) => ApplyPolicyDecision | Promise<ApplyPolicyDecision>;
  onRefusal?: (reason: RefusalCode | string, generation: number | null) => void;
  /**
   * T15: the models this application declared it can call (`models` at start). A manifest whose slot (or arm
   * override) requires a model outside it is refused locally — `model_unavailable`, the release stays unactivated,
   * nothing is fetched — and `onModelUnavailable` names the missing models for the heartbeat. Null when the
   * application declared nothing: then no slot can be refused over its model.
   */
  catalog?: readonly string[] | null;
  onModelUnavailable?: (models: readonly string[], generation: number) => void;
  /**
   * T9: called with every manifest whose envelope verifies (signature, scope,
   * generation not below the stored one) BEFORE the pass decides whether to
   * stage, hold back or ignore it — so a `disable` (Freeze) or a
   * `request_unlock` rides a manifest the runtime would otherwise leave
   * staged or unchanged. Never a manifest that failed the trust chain.
   */
  onDirectives?: (payload: Manifest["payload"]) => void;
}

export interface SyncPassOutput extends SyncPassResult {
  etag: string | null;
  edgeEtag: string | null;
  trustedRoot: RootMetadata;
  active: LoadedSlot | null;
}

/** The required models of a payload (its slots and every arm override) that the declared catalog lacks; empty when nothing was declared. */
export function requiredModelsMissing(payload: Manifest["payload"], catalog: readonly string[] | null): string[] {
  if (catalog === null) return [];
  const declared = new Set(catalog);
  const missing = new Set<string>();
  const slots = [...payload.slots, ...(payload.experiment?.arms ?? []).flatMap((arm) => arm.overrides)];
  for (const slot of slots) if (slot.modelRequired === true && !declared.has(slot.model)) missing.add(slot.model);
  return [...missing].sort();
}

export async function syncOnce(input: SyncPassInput): Promise<SyncPassOutput> {
  const now = input.now();
  let trustedRoot = input.trustedRoot;
  let edgeEtag = input.edgeEtag ?? null;
  const done = (result: SyncPassResult, active = input.active, etag = input.etag): SyncPassOutput => ({ ...result, etag, edgeEtag, trustedRoot, active });

  try {
    // A newer root document is accepted only against the one already trusted (R1–R5).
    if (input.fetchRoot) {
      const candidate = await input.fetchRoot();
      if (candidate) {
        const verdict = verifyRootMetadata({ candidate, trusted: trustedRoot, now });
        if (verdict.ok) {
          trustedRoot = candidate;
          input.store.acceptRoot(candidate);
        } else {
          input.onRefusal?.(verdict.reason, null);
        }
      }
    }

    // Idle path: the edge pointer says whether anything moved, without a Lambda on the other end. It is unsigned and
    // cacheable, so its silence is never contact (S3): the lease does not move on `pointer_unchanged`.
    if (input.edgePointerUrl && !input.skipPointer) {
      const edge = await input.client.edgePointer(input.edgePointerUrl, edgeEtag);
      if (edge.status === "not_modified") return done({ outcome: "pointer_unchanged" });
      if (edge.status === "ok") {
        edgeEtag = edge.etag;
        if (input.active && edge.pointer.generation <= input.active.generation) return done({ outcome: "pointer_unchanged" });
      }
    }

    const fetched = await input.client.manifest({ ifNoneMatch: input.etag });
    if (fetched.status === "not_modified") return done({ outcome: "unchanged" });
    if (fetched.status === "not_found") return done({ outcome: "nothing_promoted" });
    if (fetched.status === "unauthorized" || fetched.status === "forbidden") {
      input.onRefusal?.(fetched.status, null);
      return done({ outcome: "unavailable", reason: fetched.status });
    }
    if (fetched.status === "error") return done({ outcome: "unavailable", reason: `http_${fetched.httpStatus}` });

    const manifest = fetched.manifest;
    const stored = input.store.state.generation;
    const envelope = verifyManifest({ manifest, root: trustedRoot, now, scope: input.scope, storedGeneration: stored, payloads: null, countersignRoot: input.countersignRoot ?? null, ...(input.requireCountersign !== undefined ? { requireCountersign: input.requireCountersign } : {}) });
    if (!envelope.ok) {
      input.onRefusal?.(envelope.reason, manifest.payload.generation);
      return done({ outcome: "refused", reason: envelope.reason, generation: manifest.payload.generation });
    }
    // The signature verified and the generation is not a rollback: its directives stand from here on.
    input.onDirectives?.(manifest.payload);
    if (manifest.payload.generation === stored) return done({ outcome: "unchanged" }, input.active, fetched.etag);
    const heldBackBelow = input.store.state.heldBackBelow;
    if (heldBackBelow !== undefined && manifest.payload.generation <= heldBackBelow) {
      // A local rollback stepped down from this generation on purpose; only a newer one ends the hold.
      return done({ outcome: "held_back", generation: manifest.payload.generation }, input.active, fetched.etag);
    }
    // T15: a required model this runtime cannot call refuses the release here — verified, never fetched, never staged.
    const missingModels = requiredModelsMissing(manifest.payload, input.catalog ?? null);
    if (missingModels.length > 0) {
      input.onModelUnavailable?.(missingModels, manifest.payload.generation);
      input.onRefusal?.("model_unavailable", manifest.payload.generation);
      return done({ outcome: "refused", reason: "model_unavailable", generation: manifest.payload.generation }, input.active, fetched.etag);
    }

    // Fetch only what moved; the bytes already verified in the active slot are reused for unchanged hashes.
    const payloads = new Map<string, Uint8Array>();
    for (const hash of referencedPayloads(manifest.payload).keys()) {
      const held = input.active?.payloads.get(hash);
      if (held) {
        payloads.set(hash, held);
        continue;
      }
      const bytes = await input.client.payload(hash);
      if (!bytes) {
        input.onRefusal?.("payload_missing", manifest.payload.generation);
        return done({ outcome: "refused", reason: "payload_missing", generation: manifest.payload.generation });
      }
      payloads.set(hash, bytes);
    }
    const full = verifyManifest({ manifest, root: trustedRoot, now, scope: input.scope, storedGeneration: stored, payloads, countersignRoot: input.countersignRoot ?? null, ...(input.requireCountersign !== undefined ? { requireCountersign: input.requireCountersign } : {}) });
    if (!full.ok) {
      input.onRefusal?.(full.reason, manifest.payload.generation);
      return done({ outcome: "refused", reason: full.reason, generation: manifest.payload.generation });
    }

    // All-or-nothing: the current slot stays whole until the new one is complete and fsynced.
    input.store.stage({ manifest, payloads });
    const decision = await input.applyPolicy(manifest);
    if (decision === "staged") return done({ outcome: "staged", generation: manifest.payload.generation }, input.active, fetched.etag);
    // The hook activated it itself: the caller's active slot already moved; nothing here to load.
    if (decision === "activated_externally") return done({ outcome: "activated_externally", generation: manifest.payload.generation }, input.active, fetched.etag);
    const slot = input.store.activate();
    const active = input.store.load(slot, { now, root: trustedRoot, countersignRoot: input.countersignRoot ?? null, ...(input.requireCountersign !== undefined ? { requireCountersign: input.requireCountersign } : {}) });
    return done({ outcome: "activated", generation: manifest.payload.generation }, active, fetched.etag);
  } catch (error) {
    input.onRefusal?.(`network:${(error as Error).message}`, null);
    return done({ outcome: "unavailable", reason: "network" });
  }
}

export function jitteredDelayMs(baseSeconds: number, random: () => number = Math.random): number {
  // ±20 %: a fleet restarted together must not poll together.
  const jitter = (random() * 2 - 1) * 0.2;
  return Math.max(1000, Math.round(baseSeconds * 1000 * (1 + jitter)));
}
